engine/buildkit/containerimage.go

```diff
@@ -0,0 +1,178 @@
+package buildkit
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"path"
+	"strings"
+
+	"github.com/containerd/containerd/platforms"
+	bkcache "github.com/moby/buildkit/cache"
+	bkclient "github.com/moby/buildkit/client"
+	"github.com/moby/buildkit/exporter/containerimage/exptypes"
+	bkgw "github.com/moby/buildkit/frontend/gateway/client"
+	bksolverpb "github.com/moby/buildkit/solver/pb"
+	solverresult "github.com/moby/buildkit/solver/result"
+	specs "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+type ContainerExport struct {
+	Definition *bksolverpb.Definition
+	Config     specs.ImageConfig
+}
+
+func (c *Client) PublishContainerImage(
+	ctx context.Context,
+	inputByPlatform map[string]ContainerExport,
+	opts map[string]string, // TODO: make this an actual type, this leaks too much untyped buildkit api
+) (map[string]string, error) {
+	ctx, cancel, err := c.withClientCloseCancel(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer cancel()
+
+	combinedResult, err := c.getContainerResult(ctx, inputByPlatform)
+	if err != nil {
+		return nil, err
+	}
+
+	exporter, err := c.Worker.Exporter(bkclient.ExporterImage, c.SessionManager)
+	if err != nil {
+		return nil, err
+	}
+
+	expInstance, err := exporter.Resolve(ctx, opts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to resolve exporter: %s", err)
+	}
+
+	resp, descRef, err := expInstance.Export(ctx, combinedResult, c.ID())
+	if err != nil {
+		return nil, fmt.Errorf("failed to export: %s", err)
+	}
+	if descRef != nil {
+		descRef.Release()
+	}
+	return resp, nil
+}
+
+func (c *Client) ExportContainerImage(
+	ctx context.Context,
+	inputByPlatform map[string]ContainerExport,
+	destPath string,
+	opts map[string]string, // TODO: make this an actual type, this leaks too much untyped buildkit api
+) (map[string]string, error) {
+	ctx, cancel, err := c.withClientCloseCancel(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer cancel()
+
+	destPath = path.Clean(destPath)
+	if destPath == ".." || strings.HasPrefix(destPath, "../") {
+		return nil, fmt.Errorf("path %q escapes workdir; use an absolute path instead", destPath)
+	}
+
+	combinedResult, err := c.getContainerResult(ctx, inputByPlatform)
+	if err != nil {
+		return nil, err
+	}
+
+	exporterName := bkclient.ExporterDocker
+	if len(combinedResult.Refs) > 1 {
+		exporterName = bkclient.ExporterOCI
+	}
+
+	exporter, err := c.Worker.Exporter(exporterName, c.SessionManager)
+	if err != nil {
+		return nil, err
+	}
+
+	expInstance, err := exporter.Resolve(ctx, opts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to resolve exporter: %s", err)
+	}
+
+	// TODO: workaround needed until upstream fix: https://github.com/moby/buildkit/pull/4049
+	// TODO: This probably doesn't entirely work yet in the case where the combined result is still
+	// lazy and relies on other session resources to be evaluated. Fix that if merging before upstream
+	// fix in place.
```

Comment:
Is this a blocker?

Reply:
Good catch to ask about this one, I almost forgot. After thinking about it again, we can avoid that problem by just solving the result with `Evaluate: true`, which isn't even really a performance hit since that's what's gonna happen implicitly in the exporter call below anyways, so will update with that and this part of the comment removed.

Still want https://github.com/moby/buildkit/pull/4049 merged because we can delete a nice chunk of code from our codebase in exchange for that tiny upstream tweak, but not strictly necessary for correctness anymore.